src/ophyd_async/epics/pmac/_pmacTrajectory.py

```diff
@@ -0,0 +1,488 @@
+
+    def _calculate_gaps(self, chunk: Frames[PmacMotor]):
+        inds = np.argwhere(chunk.gap)
+        if len(inds) == 0:
+            return [len(chunk)]
+        else:
+            return inds
+
+
+async def ramp_up_velocity_pos(
+    motor: PmacMotor,
+    start_velocity: float,
+    end_velocity: float,
+    ramp_time: float = 0,
+    min_ramp_time: float = 0,
+):
+    # For the given motor return the displacement and time taken to get from one given
+    # velocity to another
+    acceleration_time = await motor.acceleration_time.get_value()
+    max_velocity = await motor.max_velocity.get_value()
+    accl = max_velocity / acceleration_time
+    delta_v = abs(end_velocity - start_velocity)
+    if ramp_time == 0:
+        ramp_time = delta_v / accl
+    if min_ramp_time:
+        ramp_time = max(ramp_time, min_ramp_time)
+    disp = 0.5 * (start_velocity + end_velocity) * ramp_time
+    return [disp, ramp_time]
+
+
+async def make_velocity_profile(
+    axis: PmacMotor,
+    v1: float,
+    v2: float,
+    distance: float,
+    min_time: float,
+    min_interval: float = 0.002,
+) -> vp.VelocityProfile:
+    # Create the time and velocity arrays
+    velocity_settle = 0
+    max_vel = await axis.max_velocity.get_value()
+    acc = max_vel / await axis.acceleration_time.get_value()
+    p = vp.VelocityProfile(
+        v1,
+        v2,
+        distance,
+        min_time,
+        acc,
+        max_vel,
+        velocity_settle,
+        min_interval,
+    )
+    p.get_profile()
+    return p
+
+
+async def get_gap_profile(chunk: Frames[PmacMotor], gap: int):
+    # Work out the velocity profiles of how to move to the start
+    # Turnaround can't be less than 2 ms
+    prev_point = gap - 1
+    min_turnaround = 0.002
+    min_interval = 0.002
+    time_arrays, velocity_arrays = await profile_between_points(
+        chunk,
+        gap,
+        min_turnaround,
+        min_interval,
+    )
+
+    start_positions = {}
+    for axis in chunk.axes():
+        start_positions[axis] = chunk.upper[axis][prev_point]
+
+    # Work out the Position trajectories from these profiles
+    (
+        position_array,
+        velocity_array,
+        time_array,
+    ) = await calculate_profile_from_velocities(
+        chunk, time_arrays, velocity_arrays, start_positions
+    )
+    time_profile = np.empty(len(time_array), dtype=np.float64)
+    for i in range(len(time_array)):
+        time_profile[i] = int(time_array[i] / TICK_S)
+    return position_array, velocity_array, time_profile
+
+
+async def profile_between_points(
+    chunk: Frames[PmacMotor],
+    gap: int,
+    min_time: float = 0.002,
+    min_interval: float = 0.002,
+):
+    """Make consistent time and velocity arrays for each axis
+
+    Try to create velocity profiles for all axes that all arrive at
+    'distance' in the same time period. The profiles will contain the
+    following points:-
+
+    in the following description acceleration can be -ve or +ve depending
```

Comment:
Isn't all this in https://github.com/DiamondLightSource/velocity-profile/ now?

Reply:
I don't think so. I took this from https://github.com/DiamondLightSource/pymalcolm/blob/264ec4a319092a545853f41f89ddf1d21da56504/malcolm/modules/pmac/util.py#L222 which didn't get added to velocity-profile...